### DISCUSSION TITLE
Having trouble getting this to work with external administrator user + other problems

### DISCUSSION THREAD

**Discussion by GD-Dal** [extremenetworks/ExtremeCloudIQ-APIs#5] [Q&A] [answered]
Hi.
I'm trying to build myself a platform in Postman for further development.
But I have run into trouble several places.
It seems that the "old" way to generate API Tokens are not valid anymore?
With the old way, I mean going to developer.aerohive.com -> creating user / generate client id -> log in to extremecloud and using that client ID in there. When I use this method, it fails.

The new method seems to be logging into api.extremecloudiq.com/login with a username/password and get a new token in return.
Is this correct? If so, how to set the expiry length on that one?
And this is not visible in ExtremCloudIQ instance? I can see a token being generated there, but it does not at all look like the token I get in return from api.extremecloudiq.com/login. Confusing

Anyway, this leads me to real problem:
I log into api.extremecloudiq.com/login with a with my username and password and get a new token in return.
But this user is an external user (EXT) in this VIQ ID.
So when I log in with this user, it fetches information from the original instance, which is not desirable.

Is there a way to work around this?

Thanks

**Comment by timjsmith24** [comment]
Hey,
Yes you are correct. Things are vastly different with the new API, but also things are a lot easier. 

The new method is to use the /login endpoint to generate a token. This token is valid for 24 hours and has access to all APIs the user has access to. And to you point, yes this is for only the viq that the account is a local admin on, not external. But there are API calls to address both issues. If you go to our swagger page - api.extremecloudiq.com - you can see a list of all available calls. 

1) to switch external viq (this is brand new as of a few weeks ago) - 
- run a GET call on /account/external - this will list the external accounts, each account will have a unique id (the viq id)
- run a POST call on /account/:switch?id={viq id}
This will generate a token for that user in the external viq - again a 24 hour token with full access of the user

2) generate a new token with expiration and permissions (see attached document for better details)
- run a POST call on /auth/apitoken 
- add payload 
     - set expire_time to the epoch time you would like (if removed token is good for 99 years) 
     - set permissions - you can limit the access of the token to specific sections of API calls as well as GET/POST with in those sections.
{
  "description": "Token for reading account info only",
  "expire_time": 1604737598,
  "permissions": [
    "account:r"
  ]
}

I've attached a document I wrote up prior to the latest update that included access to external accounts. Take a look at that, it covers how to get the correct permissions and generating the new tokens. I need to update this with how to switch between external accounts but hopefully this will get you started. 

If you have any issues, please feel free to reach out to me at tismith@extremenetworks.com. We can set up a zoom call to go over this if needed.

Thanks,
Tim

[A Guide to Getting Started with v2 APIs in XIQ.docx](https://github.com/extremenetworks/ExtremeCloudIQ-APIs/files/7736132/A.Guide.to.Getting.Started.with.v2.APIs.in.XIQ.docx)

**Comment by GD-Dal** [comment]
Thank you for your quick and excellent answer.
I will try this out ASAP and get back to you.

Thanks again.

**Comment by timjsmith24** [comment, marked as answer]
Something else I just realized. 

By default External user do not have api access. So a local account will need to add api access for the External Users. 
This is something that will be added to the GUI soon, but for now the PATCH /users/external/{id} call will need to be used. and the payload should include 
{
  "enable_api_access": true
}

The ID for the user can be found using the GET /users/external call.
